// solyd-keys/index.js
'use strict'
//var sodium     = require('chloride')
var crypto = require('crypto')
var sb         = require('solyd-box')
var u          = require('./util')

var isBuffer = Buffer.isBuffer

//UTILS

function clone (obj) {
  var _obj = {}
  for(var k in obj) {
    if(Object.hasOwnProperty.call(obj, k))
      _obj[k] = obj[k]
  }
  return _obj
}


function hmac(msg, key){
  return crypto.createHmac('sha256', key)
                   .update(msg)
                   .digest('hex');
}

exports.hash = u.hash

exports.getTag = u.getTag

function isObject (o) {
  return 'object' === typeof o
}

function isFunction (f) {
  return 'function' === typeof f
}

function isString(s) {
  return 'string' === typeof s
}

var curves = {}
curves.secp256k1 = require('./ektatva')

function getCurve(keys) {
  var curve = keys.curve

  if(!keys.curve && isString(keys.public))
    keys = keys.public

  if(!curve && isString(keys))
    curve = u.getTag(keys)

  if(!curves[curve]) {
    throw new Error(
      'unkown curve:' + curve +
      ' expected: '+Object.keys(curves)
    )
  }

  return curve
}

//this should return a key pair:
// {curve: curve, public: Buffer, private: Buffer}

exports.generate = function (curve, seed) {
  curve = curve || 'secp256k1'

  if(!curves[curve])
    throw new Error('unknown curve:'+curve)

  return u.keysToJSON(curves[curve].generate(seed), curve)
}

//import functions for loading/saving keys from storage
var storage = require('./storage')(exports.generate)
for(var key in storage) exports[key] = storage[key]


exports.loadOrCreate = function (filename, cb) {
  exports.load(filename, function (err, keys) {
    if(!err) return cb(null, keys)
    exports.create(filename, cb)
  })
}

exports.loadOrCreateSync = function (filename) {
  try {
    return exports.loadSync(filename)
  } catch (err) {
    return exports.createSync(filename)
  }
}


//takes a public key and a hash and returns a signature.
//(a signature must be a node buffer)

function sign (keys, msg) {
  //console.log("Sign: ", keys, msg)
  if(isString(msg))
    msg = new Buffer(msg)
  if(!isBuffer(msg))
    throw new Error('msg should be buffer')
  var curve = getCurve(keys)

  return curves[curve]
    .sign(keys.private || keys, msg)
    .toString('base64')+'.sig.'+curve

}

//takes a public key, signature, and a hash
//and returns true if the signature was valid.
function verify (keys, sig, msg) {
  if(isObject(sig))
    throw new Error('signature should be base64 string, did you mean verifyObj(public, signed_obj)')
  return curves[getCurve(keys)].verify(
    (keys.public || keys),
    u.toBuffer(sig),
    isBuffer(msg) ? msg : new Buffer(msg)
  )
}

// OTHER CRYTPO FUNCTIONS

exports.signObj = function (keys, hmac_key, obj) {
  if(!obj) obj = hmac_key, hmac_key = null
  var _obj = clone(obj)
  var b = new Buffer(JSON.stringify(_obj, null, 2))
  if(hmac_key) b = hmac(b, u.toBuffer(hmac_key))
  _obj.signature = sign(keys, b)
  return _obj
}

exports.verifyObj = function (keys, hmac_key, obj) {
  if(!obj) obj = hmac_key, hmac_key = null
  obj = clone(obj)
  var sig = obj.signature
  delete obj.signature
  var b = new Buffer(JSON.stringify(obj, null, 2))
  if(hmac_key) b = hmac(b, u.toBuffer(hmac_key))
  // console.log("V:  ", sig, b.toString(), keys)
  return verify(keys, sig, b)
}

exports.box = function (msg, recipients) {
  msg = new Buffer(JSON.stringify(msg))

  recipients = recipients.map(function (keys) {
    return (u.toBuffer(keys.public || keys)) //sodium.crypto_sign_ed25519_pk_to_curve25519
  })

  return sb.multibox(msg, recipients).toString('base64')+'.box'
}

exports.unboxKey = function (boxed, keys) {
  boxed = u.toBuffer(boxed)
  var sk = (u.toBuffer(keys.private || keys)) //sodium.crypto_sign_ed25519_sk_to_curve25519
  return sb.multibox_open_key(boxed, sk)
}

exports.unboxBody = function (boxed, key) {
  if(!key) return null
  boxed = u.toBuffer(boxed)
  key = u.toBuffer(key)
  var msg = sb.multibox_open_body(boxed, key)
  try {
    return JSON.parse(''+msg)
  } catch (_) { }
}

exports.unbox = function (boxed, keys) {
  boxed = Buffer(u.unTag(boxed),"base64")
  //console.log(">@>: ", u.unTag(boxed))
  try {
    var sk = Buffer(u.unTag(keys.private || keys),"base64")
    var msg = sb.multibox_open((boxed), sk)
    //console.log(">>>>>>>>: ",msg)
    return JSON.parse(''+msg)
  } catch (err) {
    //console.log("!>>>>>>>>: ",err)

   }
  return
}
