// solyd-keys/util.js
'use strict'
//var cl     = require('chloride')
const crypto = require('crypto')


exports.unTag =  function unTag (string) {
   //console.log("A1: ", typeof string, string)
  var i = string.split('.')
//  console.log("A: ", i)
  //var curve =  string.substring(i+1)
  return i[0]
}

exports.hash = function (data, enc) {
  data = (
    'string' === typeof data && enc == null
  ? new Buffer(data, 'binary')
  : new Buffer(data, enc)
  )
  return crypto.createHash('sha256').update(data).digest('base64')+'.sha256'
  //return cl.crypto_hash_sha256(data).toString('base64')+'.sha256'
}

exports.hasSigil = function hasSigil (s) {
  return /^(@|%|&)/.test(s)
}

function tag (key, tag) {
  //return key
  if(!tag) throw new Error('no tag for:' + key.toString('base64'))
  return key.toString('hex')+'.' + tag.replace(/^\./, '')
}

exports.keysToJSON = function keysToJSON(keys, curve) {
  curve = (keys.curve || curve)

  var pub = tag(keys.public.toString('utf8'), curve)
  return {
    curve: curve,
    public: pub,
    private: keys.private ? tag(keys.private.toString(), curve) : undefined,
    id: '@'+(curve === 'secp256k1' ? pub : exports.hash(pub))
  }
}

exports.getTag = function getTag (string) {
  var i = string.indexOf('.')
  return string.substring(i+1)
}

exports.toBuffer = function (buf) {
  if(buf == null) return buf
  if(Buffer.isBuffer(buf)) return buf
  var i = buf.indexOf('.')
  return new Buffer(buf, 'hex')
  //var start = (exports.hasSigil(buf)) ? 1 : 0
  //return new Buffer(buf.substring(start, ~i ? i : buf.length), 'hex')
}
